ARROW-6101: [Rust] [DataFusion] Parallel execution of physical query plan

andygrove opened the pull request:
This PR implements parallel query execution, with the ability to create a physical query plan from a logical plan, and call `ExecutionContext.collect()` to execute the plan and collect the results into a single vector.

This currently only supports trivial projections against partitioned CSV files.

I will create separate PRs to add the other operators (selection, limit, aggregation, etc).

nevi-me commented:
Hi @andygrove, I'll catch up on this and other PRs during the week

paddyhoran commented on rust/datafusion/src/execution/context.rs:
```diff
@@ -210,6 +216,103 @@ impl ExecutionContext {
+                ..
+            } => match (*self.datasources).borrow().get(table_name) {
+                Some(provider) => {
+                    let partitions = provider.scan(projection, 16 * 1024)?;
```
What's the rationale for hard coding `16 * 1024`?

andygrove replied:
I've made this a parameter now.

paddyhoran commented on rust/datafusion/src/execution/context.rs:
```diff
@@ -210,6 +216,103 @@ impl ExecutionContext {
+                            "Table provider returned no partitions".to_string(),
+                        ))
+                    } else {
+                        let partition = partitions[0].lock().unwrap();
```
Can we avoid the `unwrap` here?

andygrove replied:
In this case, I think we want the unwrap. From https://doc.rust-lang.org/std/sync/struct.Mutex.html:

```
Most usage of a mutex will simply unwrap() these results, propagating panics among threads to ensure that a possibly invalid invariant is not witnessed.
```

These queries are now happening on threads, so although a panic here would fail the query, it would not fail the main thread that launched the thread(s) to execute the query.

paddyhoran commented on rust/datafusion/src/execution/context.rs:
```diff
@@ -210,6 +216,103 @@ impl ExecutionContext {
+                        Ok(Arc::new(exec))
+                    }
+                }
+                _ => panic!(),
```
`panic` should provide some error message

andygrove replied:
Oops. That one was not intentional ... fixed.

paddyhoran commented on rust/datafusion/src/execution/context.rs:
```diff
@@ -210,6 +216,103 @@ impl ExecutionContext {
+            LogicalPlan::Projection { input, expr, .. } => {
+                let input = self.create_physical_plan(input)?;
+                let input_schema = input.as_ref().schema().clone();
+                //let me = self;
```
Remove comment.

paddyhoran commented on rust/datafusion/src/execution/context.rs:
```diff
@@ -210,6 +216,103 @@ impl ExecutionContext {
+            .map(|p| {
+                let p = p.clone();
+                thread::spawn(move || {
+                    let it = p.execute().unwrap();
```
Again do we need to `unwrap` here two lines in a row?

andygrove replied:
Fixed.

paddyhoran commented on rust/datafusion/src/execution/context.rs:
```diff
@@ -210,6 +216,103 @@ impl ExecutionContext {
+        // combine the results from each thread
+        let mut combined_results: Vec<RecordBatch> = vec![];
+        for thread in threads {
+            let result = thread.join().unwrap();
```
`unwrap` again

andygrove replied:
Fixed

andygrove commented:
Thanks @paddyhoran. I have addressed the comments, except for the two uses of `unwrap` when locking mutexes. See my comment above related to that. Let me know what you think.

andygrove commented:
Travis failure unrelated, so going ahead with merge.

```+sudo apt-get update
Hit:1 http://apt.postgresql.org/pub/repos/apt xenial-pgdg InRelease
Hit:2 http://archive.ubuntu.com/ubuntu xenial InRelease
Hit:3 http://archive.ubuntu.com/ubuntu xenial-updates InRelease
Hit:4 http://security.ubuntu.com/ubuntu xenial-security InRelease
Hit:5 http://archive.ubuntu.com/ubuntu xenial-backports InRelease
Hit:6 https://apt.llvm.org/xenial llvm-toolchain-xenial-7 InRelease
Get:7 https://packages.microsoft.com/ubuntu/16.04/prod xenial InRelease [3,226 B]
Get:8 https://packages.microsoft.com/ubuntu/16.04/prod xenial/main amd64 Packages [108 kB]
Err:8 https://packages.microsoft.com/ubuntu/16.04/prod xenial/main amd64 Packages
  Hash Sum mismatch
Fetched 112 kB in 0s (132 kB/s)
Reading package lists... Done
E: Failed to fetch https://packages.microsoft.com/ubuntu/16.04/prod/dists/xenial/main/binary-amd64/Packages.gz  Hash Sum mismatch
E: Some index files failed to download. They have been ignored, or old ones used instead.
The command "$TRAVIS_BUILD_DIR/ci/travis_release_test.sh" exited with 100.```
